ibis/expr/types/joins.py

```diff
@@ -81,15 +81,15 @@ def dereference_sides(left, right, deref_left, deref_right):
     return left, right
 
 
-def dereference_binop(pred, deref_left, deref_right):
+def dereference_comparison_op(pred, deref_left, deref_right):
     left, right = dereference_sides(pred.left, pred.right, deref_left, deref_right)
     return pred.copy(left=left, right=right)
 
 
 def dereference_value(pred, deref_left, deref_right):
     deref_both = {**deref_left, **deref_right}
-    if isinstance(pred, ops.Binary) and pred.left.relations == pred.right.relations:
-        return dereference_binop(pred, deref_left, deref_right)
+    if isinstance(pred, ops.Comparison) and pred.left.relations == pred.right.relations:
```

Comment:
The point of this branch is to separately try to dereference each side of the predicate. This basically allows us to write

```py
t1 = ...

t1.join(t1, t1.a == t1.a)
```

Reply:
But this indeed should be applied only to comparisons not e.g. `&` `|` ...

Reply:
Thanks for the explanation!

Reply:
Otherwise it tries to attempt to dereference all the values in the predicate to either sides. The test case works because of `t.view()` creates two distinct tables so the references in the predicates are not ambiguous.